src/generic/gridsel.cpp

```diff
@@ -398,517 +288,262 @@ void wxGridSelection::SelectBlock( int topRow, int leftCol,
-            if ( row == sel.GetRow() && col == sel.GetCol() )
-            {
-                wxGridCellCoords coords = m_cellSelection[n];
-                m_cellSelection.RemoveAt(n);
-                if ( !m_grid->GetBatchCount() )
-                {
-                    m_grid->RefreshBlock(coords, coords);
-                }
-
-                // Send event
-                wxGridRangeSelectEvent gridEvt( m_grid->GetId(),
-                                                wxEVT_GRID_RANGE_SELECT,
-                                                m_grid,
-                                                wxGridCellCoords( row, col ),
-                                                wxGridCellCoords( row, col ),
-                                                false,
-                                                kbd );
-                m_grid->GetEventHandler()->ProcessEvent( gridEvt );
-
-                return;
-            }
-        }
-    }
-
-    // The most difficult case: The cell is member of one or even several
-    // blocks. Split each such block in up to 4 new parts, that don't
-    // contain the cell to be selected, like this:
+    // If the selected block intersects with the deselection block, split it
+    // in up to 4 new parts, that don't contain the block to be selected, like
+    // this (for rows):
     // |---------------------------|
     // |                           |
     // |           part 1          |
     // |                           |
     // |---------------------------|
-    // |   part 3   |x|   part 4   |
+    // | part 3 |    x    | part 4 |
     // |---------------------------|
     // |                           |
     // |           part 2          |
     // |                           |
     // |---------------------------|
-    //   (The x marks the newly deselected cell).
-    // Note: in row selection mode, we only need part1 and part2;
-    //       in column selection mode, we only need part 3 and part4,
-    //          which are expanded to whole columns automatically!
+    // And for columns:
+    // |---------------------------|
+    // |        |         |        |
+    // |        | part 3  |        |
+    // |        |         |        |
+    // |        |---------|        |
+    // | part 1 |    x    | part 2 |
+    // |        |---------|        |
+    // |        |         |        |
+    // |        | part 4  |        |
+    // |        |         |        |
+    // |---------------------------|
+    //   (The x marks the newly deselected block).
+    // Note: in row/column selection mode, we only need part1 and part2
+
+    // Blocks to refresh.
+    wxVectorGridBlockCoords refreshBlocks;
+    // Add the deselected block.
+    refreshBlocks.push_back(wxGridBlockCoords(topRow, leftCol,
+                                              bottomRow, rightCol));
+
+    count = m_selection.size();
+    for ( n = 0; n < count; n++ )
+    {
```

Comment:
If the body of this loop could be refactored in another `wxGridBlockCoords` method or a private function (i.e. a function in a private header) that could be unit tested directly, it would help a lot as it's really difficult to be sure that there are no bugs here right now.